sdk/src/azure/iot/az_iot_adu_client.c

```diff
@@ -0,0 +1,872 @@
+      {
+        update_request->update_manifest_signature = ref_json_reader->token.slice;
+      }
+    }
+    else if (az_json_token_is_text_equal(
+                 &ref_json_reader->token,
+                 AZ_SPAN_FROM_STR(AZ_IOT_ADU_CLIENT_AGENT_PROPERTY_NAME_FILEURLS)))
+    {
+      _az_RETURN_IF_FAILED(az_json_reader_next_token(ref_json_reader));
+      if (ref_json_reader->token.kind != AZ_JSON_TOKEN_NULL)
+      {
+        RETURN_IF_JSON_TOKEN_NOT_TYPE(ref_json_reader, AZ_JSON_TOKEN_BEGIN_OBJECT);
+        _az_RETURN_IF_FAILED(az_json_reader_next_token(ref_json_reader));
+
+        while (ref_json_reader->token.kind != AZ_JSON_TOKEN_END_OBJECT)
+        {
+          RETURN_IF_JSON_TOKEN_NOT_TYPE(ref_json_reader, AZ_JSON_TOKEN_PROPERTY_NAME);
+
+          update_request->file_urls[update_request->file_urls_count].id
+              = ref_json_reader->token.slice;
+
+          _az_RETURN_IF_FAILED(az_json_reader_next_token(ref_json_reader));
+          if (ref_json_reader->token.kind != AZ_JSON_TOKEN_NULL)
+          {
+            update_request->file_urls[update_request->file_urls_count].url
+                = ref_json_reader->token.slice;
+
+            update_request->file_urls_count++;
+          }
+
+          _az_RETURN_IF_FAILED(az_json_reader_next_token(ref_json_reader));
+        }
+      }
+    }
+
+    _az_RETURN_IF_FAILED(az_json_reader_next_token(ref_json_reader));
+  }
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result az_iot_adu_client_get_service_properties_response(
+    az_iot_adu_client* client,
+    int32_t version,
+    int32_t status,
+    az_json_writer* ref_json_writer)
+{
+  _az_PRECONDITION_NOT_NULL(client);
+  _az_PRECONDITION_NOT_NULL(ref_json_writer);
+
+  (void)client;
+
+  // Component and response status
+  _az_RETURN_IF_FAILED(az_json_writer_append_begin_object(ref_json_writer));
+  _az_RETURN_IF_FAILED(az_iot_hub_client_properties_writer_begin_component(
+      NULL, ref_json_writer, AZ_SPAN_FROM_STR(AZ_IOT_ADU_CLIENT_AGENT_COMPONENT_NAME)));
+  _az_RETURN_IF_FAILED(az_iot_hub_client_properties_writer_begin_response_status(
+      NULL,
+      ref_json_writer,
+      AZ_SPAN_FROM_STR(AZ_IOT_ADU_CLIENT_AGENT_PROPERTY_NAME_SERVICE),
+      status,
+      version,
+      AZ_SPAN_EMPTY));
+
+  // It is not necessary to send the properties back in the acknowledgement.
+  // We opt not to send them to reduce the size of the payload.
+  _az_RETURN_IF_FAILED(az_json_writer_append_begin_object(ref_json_writer));
+  _az_RETURN_IF_FAILED(az_json_writer_append_end_object(ref_json_writer));
+
+  _az_RETURN_IF_FAILED(
+      az_iot_hub_client_properties_writer_end_response_status(NULL, ref_json_writer));
+  _az_RETURN_IF_FAILED(az_iot_hub_client_properties_writer_end_component(NULL, ref_json_writer));
+  _az_RETURN_IF_FAILED(az_json_writer_append_end_object(ref_json_writer));
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result az_iot_adu_client_parse_update_manifest(
```

Comment:
Please split this function into smaller (inline?) functions to reduce complexity and increase readability.

Reply:
For time, not going to do this right now but I'll leave this open so that we may do this later.